File: funcscript/funcs/misc/LogFunction.cs
using funcscript.core;
using funcscript.model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace funcscript.funcs.misc
{
    public class LogFunction : IFsFunction
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Infix;

        public string Symbol => "log";

        public int Precidence => 0;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count == 0)
                throw new error.EvaluationTimeException($"{this.Symbol} function: {this.ParName(0)} expected");

            var tag = pars.Count > 1 ? $"({pars.GetParameter(parent, 1).ToString()})" : "";
            var output = pars.Count > 2 ? (pars.GetParameter(parent, 2) is bool ? (bool)pars.GetParameter(parent, 2) : false) : true;
            Console.WriteLine($"FuncScript: Evaluating {tag}");
            try
            {
                var res = pars.GetParameter(parent, 0);
                if (output)
                {
                    Console.WriteLine($"FuncScript: Result{tag}:\n{(res == null ? "<null>" : res.ToString())}");
                    Console.WriteLine($"End Result {tag}");
                }
                else
                    Console.WriteLine($"Done {tag}");
                return res;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"FuncScript: Error evaluating {tag}");
                var thisEx = ex;
                while (thisEx != null)
                {
                    Console.WriteLine(thisEx.Message);
                    Console.WriteLine(thisEx.StackTrace);
                    thisEx = thisEx.InnerException;
                }
                throw;
            }
        }
        public string ParName(int index)
        {
            switch(index)
            {
                case 0: return "expression";
                case 1: return "tag";
                case 2: return "output";
                default:return null;
            }
        }
    }
}

File: js-port/fs-js/src/funcs/misc/log-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');

class LogFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'log';
    this.callType = CallType.Infix;
  }

  get maxParameters() {
    return -1;
  }

  evaluate(provider, parameters) {
    if (parameters.count === 0) {
      return helpers.makeError(helpers.FsError.ERROR_PARAMETER_COUNT_MISMATCH, `${this.symbol}: expression expected`);
    }

    const expression = parameters.getParameter(provider, 0);
    const tag = parameters.count > 1 ? String(parameters.getParameter(provider, 1) ?? '') : '';
    const outputResult = parameters.count > 2 ? helpers.ensureTyped(parameters.getParameter(provider, 2)) : null;
    const output = outputResult ? helpers.valueOf(outputResult) : true;

    const messageTag = tag ? `(${tag})` : '';
    console.log(`FuncScript: Evaluating ${messageTag}`);
    try {
      const result = helpers.ensureTyped(expression);
      if (output) {
        console.log(`FuncScript: Result ${messageTag}:`);
        console.log(helpers.valueOf(result));
      } else {
        console.log(`FuncScript: Done ${messageTag}`);
      }
      return result;
    } catch (error) {
      console.error(`FuncScript: Error evaluating ${messageTag}`);
      console.error(error);
      throw error;
    }
  }
}

module.exports = {
  LogFunction
};
